arangod/Network/NetworkFeature.cpp

```diff
@@ -238,4 +264,59 @@ void NetworkFeature::trackForwardedRequest() {
   ++_forwardedRequests;
 }
 
+std::size_t NetworkFeature::requestsInFlight() const {
+  return _requestsInFlight.load();
+}
+
+bool NetworkFeature::isCongested() const {
+  return _requestsInFlight.load() >= (_maxInFlight * ::CongestionRatio);
+}
+
+bool NetworkFeature::isSaturated() const {
+  return _requestsInFlight.load() >= _maxInFlight;
+}
+
+void NetworkFeature::sendRequest(network::ConnectionPool& pool,
+                                 network::RequestOptions const& options,
+                                 std::string const& endpoint,
+                                 std::unique_ptr<fuerte::Request>&& req,
+                                 RequestCallback&& cb) {
+  prepareRequest(pool, req);
+  auto conn = pool.leaseConnection(endpoint);
+  conn->sendRequest(std::move(req),
+                    [this, &pool,
+                     cb = std::move(cb)](fuerte::Error err,
+                                         std::unique_ptr<fuerte::Request> req,
+                                         std::unique_ptr<fuerte::Response> res) {
+                      finishRequest(pool, err, req, res);
+                      cb(err, std::move(req), std::move(res));
+                    });
+}
+
+void NetworkFeature::prepareRequest(network::ConnectionPool const& pool,
+                                    std::unique_ptr<fuerte::Request>& req) {
+  _requestsInFlight += 1;
+  if (req) {
+    req->timestamp(std::chrono::steady_clock::now());
+  }
+}
+void NetworkFeature::finishRequest(network::ConnectionPool const& pool, fuerte::Error err,
+                                   std::unique_ptr<fuerte::Request> const& req,
+                                   std::unique_ptr<fuerte::Response>& res) {
+  _requestsInFlight -= 1;
+  if (err == fuerte::Error::RequestTimeout) {
+    _requestTimeouts.count();
+  } else if (req && res) {
+    res->timestamp(std::chrono::steady_clock::now());
+    std::chrono::milliseconds duration =
+        std::chrono::duration_cast<std::chrono::milliseconds>(res->timestamp() -
+                                                              req->timestamp());
+    std::chrono::milliseconds timeout = req->timeout();
+    double percentage = std::clamp(100.0 * static_cast<double>(duration.count()) /
+                                       static_cast<double>(timeout.count()),
+                                   0.0, 100.0);
```

Comment:
Is it possible that `timeout` somehow is 0 here? In this case we could get a SIGFPE.

Reply:
If it is, I think we have bigger problems.